main.go

```diff
@@ -0,0 +1,229 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/c3os-io/c3os/provider-kubeadm/pkg/config"
+	"gopkg.in/yaml.v2"
+	"path/filepath"
+	"text/template"
+
+	"github.com/c3os-io/c3os/sdk/clusterplugin"
+	yip "github.com/mudler/yip/pkg/schema"
+	"github.com/sirupsen/logrus"
+)
+
+var configurationPath = "/opt/kubeadm"
+
+func clusterProvider(cluster clusterplugin.Cluster) yip.YipConfig {
+
+	// TODO(kpiyush17) add cluster.Options handling, expecting the options to match KubeadmConfig spec
+
+	var stages []yip.Stage
+
+	preStage := yip.Stage{
+		Systemctl: yip.Systemctl{
+			Enable: []string{"kubelet"},
+			Start:  []string{"containerd"},
+		},
+		Commands: []string{"sudo sysctl --system"},
+	}
+
+	stages = append(stages, preStage)
+
+	if cluster.Role == clusterplugin.RoleInit {
+		stages = append(stages, getInitYipStages(cluster)...)
+	} else if (cluster.Role == clusterplugin.RoleControlPlane) || (cluster.Role == clusterplugin.RoleWorker) {
+		stages = append(stages, getJoinYipStages(cluster)...)
+	}
+
+	cfg := yip.YipConfig{
+		Name: "Kubeadm C3OS Cluster Provider",
+		Stages: map[string][]yip.Stage{
+			"boot.before": stages,
+		},
+	}
+
+	return cfg
+}
+
+func main() {
```

Comment:
can we move main to the top of the functions, they should be listed in the order they are used